app/controllers/incomeSources/add/IncomeSourceAddedController.scala

```diff
@@ -0,0 +1,164 @@
+    implicit request =>
+      implicit user =>
+        getMtdItUserWithIncomeSources(incomeSourceDetailsService) flatMap {
+          implicit mtdItUser =>
+            handleRequest(isAgent = true, incomeSourceId, incomeSourceType)
+        }
+  }
+
+  private def getBackUrl(incomeSourceId: String, isAgent: Boolean, incomeSourceType: IncomeSourceType): String = {
+    val baseRoute = incomeSourceType match {
+      case SelfEmployment =>
+        if (isAgent) controllers.incomeSources.add.routes.BusinessReportingMethodController.showAgent _ else
+          controllers.incomeSources.add.routes.BusinessReportingMethodController.show _
+      case UkProperty =>
+        if (isAgent) controllers.incomeSources.add.routes.UKPropertyReportingMethodController.showAgent _ else
+          controllers.incomeSources.add.routes.UKPropertyReportingMethodController.show _
+      case ForeignProperty =>
+        if (isAgent) controllers.incomeSources.add.routes.ForeignPropertyReportingMethodController.showAgent _ else
+          controllers.incomeSources.add.routes.ForeignPropertyReportingMethodController.show _
+    }
+    baseRoute(incomeSourceId).url
+  }
+
+  def getIncomeSourceData(incomeSourceType: IncomeSourceType, incomeSourceId: String)(implicit user: MtdItUser[_]): Option[(LocalDate, Option[String])] = {
+    incomeSourceType match {
+      case SelfEmployment =>
+        user.incomeSources.businesses
+          .find(_.incomeSourceId.equals(incomeSourceId))
+          .flatMap { addedBusiness =>
+            for {
+              businessName <- addedBusiness.tradingName
+              startDate <- addedBusiness.tradingStartDate
+            } yield (startDate, Some(businessName))
+          }
+      case UkProperty =>
+        for {
+          newlyAddedProperty <- user.incomeSources.properties.find(incomeSource =>
+            incomeSource.incomeSourceId.equals(incomeSourceId) && incomeSource.isUkProperty
+          )
+          startDate <- newlyAddedProperty.tradingStartDate
+        } yield (startDate, None)
+      case ForeignProperty =>
+        for {
+          newlyAddedProperty <- user.incomeSources.properties.find(incomeSource =>
+            incomeSource.incomeSourceId.equals(incomeSourceId) && incomeSource.isForeignProperty
+          )
+          startDate <- newlyAddedProperty.tradingStartDate
+        } yield (startDate, None)
+    }
+  }
+
+  private def handleRequest(isAgent: Boolean, incomeSourceId: String, incomeSourceType: IncomeSourceType)(implicit user: MtdItUser[_], ec: ExecutionContext): Future[Result] = {
+    withIncomeSourcesFS {
+      val backUrl = getBackUrl(incomeSourceId, isAgent, incomeSourceType)
+      getIncomeSourceData(incomeSourceType, incomeSourceId) match {
+        case Some(incomeSourceData) =>
+          val showPreviousTaxYears: Boolean = incomeSourceData._1.isBefore(dateService.getCurrentTaxYearStart())
```

Comment:
avoid resolve Tuple type in the ways above/below: =>
._1 or ._2
you can alway resolve these values in the variable with semantical meaning defined by the actual variable name